README.md
# ![ttl-mem-cache](logo.png)

A in memory time to live cache with streaming support.

[![GitHub Actions status](https://github.com/trygve-lie/ttl-mem-cache/workflows/test/badge.svg)](https://github.com/trygve-lie/ttl-mem-cache/actions?query=workflow%3A%22test%22)
[![Known Vulnerabilities](https://snyk.io/test/github/trygve-lie/ttl-mem-cache/badge.svg?targetFile=package.json&style=flat-square)](https://snyk.io/test/github/trygve-lie/ttl-mem-cache?targetFile=package.json)


## Installation

```bash
$ npm install ttl-mem-cache
```


## Example

Set an object and retrieve it.

```js
const Cache = require('ttl-mem-cache');

const cache = new Cache();

cache.set('a', {foo: 'bar'});
const obj = cache.get('a'); // returns {foo: 'bar'}
```


## Description

This is a in memory time to live key/value cache with streaming support. Items are
not pro-actively pruned out as they age but expires when they are too old when touched.
In other words; this module does not use `setTimeout()` or simmilar methods internally.

There is no restrictions on the values stored in the cache and there is no maximum limit
of the amount of items in the cache. Whan that is said, the intention of this module is
to act as a simple key/value cache where one need to cache small to medium amounts of
data.


## Constructor

Create a new Cache instance.

```js
const Cache = require('ttl-mem-cache');
const cache = new Cache(options);
```

### options (optional)

An Object containing misc configuration. The following values can be provided:

 * ttl - `Number` - Default time to live in milliseconds all items in the cache should be cached before expiering.
 * stale - `Boolean` - If expired items in cache should be returned when pruned from the cache. Default: `false`.
 * changelog - `Boolean` - If emitted `set` event and stream should contain both old and new value. Default: `false`.
 * id - `String` - Give the instanse a unique identifier. Default: `hash`

If an option Object with a `ttl` is not provided all items in the cache will by default
cached for 5 minutes before they expire.

Items can be cached forever by setting `ttl` to `Infinity`. Items cached forever can
be overwritten and manually deleted.

Pruning of items from the cache happend when they are touched by one of the methods
for retrieving (`.get()` and `.entries()`) items from the cache. By default pruning
happens before the method returns a value so if an item have expired, `null` will be
returned for expired items. By setting `stale` to `true`, these methods will return
the pruned item(s) before they are removed from the cache.

Internally the cache has a unique ID created each time its instantiated. This ID is
used to tell the origin of an cached item when streaming. The `id` will override
this generated ID. When using this, be carefull to not provide the same ID to multiple
instances of the cache.

The Cache instance inherit from Duplex Stream. Due to this the instance also take all
config parameters which the Duplex Stream does. Please see the [documentation of Duplex Streams](https://nodejs.org/api/stream.html#stream_duplex_and_transform_streams)
for further documentation.


## API

The Cache instance have the following API:


### .set(key, value, ttl)

Set an item in the cache on a given key.

```js
const cache = new Cache();
cache.set('a', {foo: 'bar'});
cache.set('b', {foo: 'xyz'}, 20 * 60 * 1000);
```

This method take the following arguments:

 * key - An unique key the value should be stored on in the cache. Required.
 * value - The value to store on the key in the cache. Required.
 * ttl - Time to live before the item should expire. Uses default if not given. Optional.

An item can be cached forever by setting `ttl` to `Infinity`.


### .get(key)

Get an item on a given key from the cache.

```js
const cache = new Cache();
cache.set('a', {foo: 'bar'});
const obj = cache.get('a'); // returns {foo: 'bar'}
```

This method take the following arguments:

 * key - The unique key for the item in the cache. Required.

Triggering `.get()` will check the expire on the item. If the item is older than
the time to live set on it, the item will be removed from the cache and this method
will return `null` unless `stale` is set to `true` on the constructor. Then the
expired item will be returned before its removed from the cache.


### .del(key)

Explicitly delete an item on a given key in the cache.

```js
const cache = new Cache();
cache.set('a', {foo: 'bar'});
cache.del('a');
```

This method take the following arguments:

 * key - The unique key for the item in the cache. Required.


### .entries(mutator)

Get all items in the cache. Returns an Array with all items.

```js
const cache = new Cache();
cache.set('a', {foo: 'bar'});
cache.set('b', {foo: 'xyz'});
const all = cache.entries(); // returns [{foo: 'bar'}, {foo: 'xyz'}]
```

Triggering `.entries()` will check the expire on the items. If an item is older than
the time to live set on it, the item will be removed from the cache and it will not be
included in the returned value of this methid unless `stale` is set to `true` on the
constructor. Then the expired item will be included before its removed from the cache.

This method take the following arguments:

 * mutator - A function for mutating the items returned. Optional.

The mutator attribute can be used to change the structure of the returned items. It
takes a function which will be called with an Object with the `key` and `value`. This
function must return a value which then will be the value of the item in the returned
output.

```js
const cache = new Cache();
cache.set('a', {foo: 'bar'});
cache.set('b', {foo: 'xyz'});
const all = cache.entries((item) => {
    return item.value.foo;
}); // returns ['bar', 'xyz']
```

The advantage of using the mutator if one want to mutate the items is that the mutator
is applied to each item in the same process as the expire is checked.


### .prune()

Iterates over all items in the cache and proactively prunes expired items.


### .clear()

Clears the entire cache. All items will be deleted.


### .dump()

Returns an Array of all items in the cache ready to be used by `.load()`.


### .load(dump)

Loads an Array of items, provided by `.dump()`, into the cache.

This method take the following arguments:

 * dump - Array of items to be imported.

If any of the items in the loaded Array contains a key which already are in
the cache the entry in the cache will be overwritten.

If any of the entries in the loaded Array are not compatible with the format
which `.dump()` exports, they will not be inserted into the cache.

Returns and Array with the keys which was inserted into the cache.


### .length()

The number of items in the cache.


## Events

The Cache instance inherit from Duplex Stream. Due to this the instance emits all the
events which Duplex Stream does when the streaming feature is used. Please see the
[documentation of Duplex Streams](https://nodejs.org/api/stream.html#stream_duplex_and_transform_streams)
for further documentation.

In addition to this, the following events are emitted:


### set

When an item is set in the cache. Emits an Object with the `key` and `value` of the item.

```js
const cache = new Cache();
cache.on('set', (key, item) => {
    console.log(key, item);  // outputs: "a, {foo: 'bar'}"
});
cache.set('a', {foo: 'bar'});
```

If `changefeed` is set to be `true` on the constructor, the emitted Object will hold both
old and new value for the key. See "changelog" for further info.

### dispose

When an item is disposed (deleted) from the cache. Emits the `key` of the item and the item itself.

```js
const cache = new Cache();
cache.on('dispose', (key, item) => {
    console.log(key, item);  // outputs: "a, {foo: 'bar'}"
});
cache.set('a', {foo: 'bar'});
cache.del('a');
```

### clear

When the cache is cleared.


## Streams

The Cache instance is a [Duplex Stream](https://nodejs.org/api/stream.html#stream_duplex_and_transform_streams). One can stream
items in and out of the cache.

Example of streaming into the cache:

```js
const cache = new Cache();
const source = new SomeReadableStream();  // a source streaming an item on key 'a' into the cache

source.pipe(cache);

cache.get('a');  // returns the item for key 'a'
```

Example of streaming out of the cache:

```js
const cache = new Cache();
const dest = new SomeWritableStream();  // a destination stream recieving items from the cache

cache.pipe(dest);

cache.set('a', {foo: 'bar'});  // pushes {foo: 'bar'} onto the writable stream
```

Example of streaming through the cache (all items streamed through will be kept in the cache):

```js
const cache = new Cache();
const source = new SomeReadableStream();  // a source streaming items into the cache
const dest = new SomeWritableStream();  // a destination stream recieving items from the cache

source.pipe(cache).pipe(dest);

cache.entries();  // returns all the items streamed through the cache
```


### Linking caches

With the stream API its possible to link caches together and distribute cached items between them.

```js
const Cache = require('../');

const cacheA = new Cache();
const cacheB = new Cache();
const cacheC = new Cache();

// Link all caches together
cacheA.pipe(cacheB).pipe(cacheC).pipe(cacheA);

// Set a value in cache C
cacheC.set('foo', 'bar');

// Retrieve the same value from all caches
console.log(cacheA.get('foo'), cacheB.get('foo'), cacheC.get('foo'));

// Delete the value in cache A
cacheA.del('foo');

// The value is deleted from all caches
console.log(cacheA.get('foo'), cacheB.get('foo'), cacheC.get('foo'));
```


### Streaming Object type

When using the Stream API to write to and read from the cache an [Entry Object](https://github.com/trygve-lie/ttl-mem-cache/blob/master/lib/entry.js)
or an Object of simmilar character is used.

The [Entry Object](https://github.com/trygve-lie/ttl-mem-cache/blob/master/lib/entry.js)
looks like this:

```js
{
    key: 'item key',
    value: 'item value',
    origin: 'cache instance ID',
    ttl: 'time to live',
    expires: 'time stamp'
}
```

The Stream API will always emit a full Entry Object. When writing to the Stream API one can provide
a full Entry Object, but a simmilar Object will also be accepted. Only `key` and `value` are required
properties when writing to the Stream API.

Iow; the following Object will be accepted by the Stream API:

```js
{
    key: 'foo',
    value: 'bar'
}
```

`key` defines what key the value of `value` should be stored on in the cache. `key` is required
and if not provided the stream will emit an error.

When writing to the cache and a Object with `value` with a value is provided, it will be stored in
the cache on the provided `key`. If `value` is not provided or `null` or `undefined` on the
Object when writing to the cache, any item with a matching `key` in the cache will be deleted.

When the stream emits objects each object will have a `origin` key. The value is the unique ID of
the instance the object first was emitted on the stream. If the construcor was given a value for
the `id` argument, this will be used as `origin` value.

`ttl` is how long the item is set to live in the cache.

`expires` is the calculated timestamp for when the item should expire from the cache. This is
calculated when an item is set in the cache. If a value for `expires` is provided when writing
to the Stream API this value will be set in the cache.

If the items you want to store in the cache does not fit your data type, its recommended to use
a [Transform Stream](https://nodejs.org/api/stream.html#stream_implementing_a_transform_stream)
to transform it into the supported Object type.

Example:

```js
const cache = new Cache();
const source = new SomeReadableStream();  // contains Objects like this {id: 'a', item: 'bar'}

const convert = new stream.Transform({
    objectMode: true,
    transform(obj, encoding, callback) {
        this.push({
            key: obj.id,
            value: obj.item
        });
        callback();
    }
});

source.pipe(convert).pipe(cache);
```

The [Entry Object](https://github.com/trygve-lie/ttl-mem-cache/blob/master/lib/entry.js) does
implement `.toPrimitive` where a stringified JSON representation of the Object will be returned
when call to the Entry Object with a String hint is done.

This can be used when one want to convert the Entry Object into a [Buffer](https://nodejs.org/api/buffer.html).

```js
const cache = new Cache();
const dest = new SomeWritableStream();  // Some destination supporting Buffers

const convert = new stream.Transform({
    writableObjectMode: true,
    readableObjectMode: false,
    transform(obj, encoding, callback) {
        const buff = Buffer.from(obj);  // Convert Entry Object to a Buffer
        this.push(buff);
        callback();
    }
});

cache.pipe(convert).pipe(dest);
```

### Object Mode

By default the stream are in object mode. Its also supported to run the stream in non-object
mode. Object mode can be turned off by setting `objectMode` to `false` as an option to the
constructor.

When using non-object mode the stream will emit `Buffers` containing a stringified JSON
representaion of the Entry Object.


## Changelog

If the attribute `changelog` is set to `true` on the constructor, the `set` events will emit an
object holding both old and new values for the key.

The emitted object looks like this:

```js
{
    oldVal: 'foo',
    newVal: 'bar'
}
```

Example:

```js
const Cache = require('ttl-mem-cache');

const cache = new Cache({ changefeed: true });
cache.on('set', (key, item) => {
    // item will be in the format above
});
cache.set('a', 'foo');
cache.set('a', 'bar');
```

If a key does not hold a value in cache before, `oldVal` will be `null`.
If a key hold a value which has expired and `stale` is `false`, `oldVal` will be `null`.
If a key hold a value which has expired and `stale` is `true`, `oldVal` will be the old value.


## error handling

This module does not handle errors for you, so you must handle errors on
whatever streams you pipe into this module. This is a general rule when
programming with node.js streams: always handle errors on each and every stream.

We recommend using [`end-of-stream`](https://npmjs.org/end-of-stream) or [`pump`](https://npmjs.org/pump)
for writing error tolerant stream code.


## License

The MIT License (MIT)

Copyright (c) 2017 - Trygve Lie - post@trygve-lie.com

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
## [4.1.3](https://github.com/trygve-lie/ttl-mem-cache/compare/v4.1.2...v4.1.3) (2023-05-02)


### Bug Fixes

* Bump version ([5e45f21](https://github.com/trygve-lie/ttl-mem-cache/commit/5e45f21cfd3cf734f990df5c38798cdca0876812))

## [4.1.2](https://github.com/trygve-lie/ttl-mem-cache/compare/v4.1.1...v4.1.2) (2023-05-02)


### Bug Fixes

* Bump version ([ec6b01d](https://github.com/trygve-lie/ttl-mem-cache/commit/ec6b01d838ce3f639bd48b5ebd20743fb03dfffa))

## [4.1.1](https://github.com/trygve-lie/ttl-mem-cache/compare/v4.1.0...v4.1.1) (2023-05-02)


### Bug Fixes

* Do not reference this in module ([#107](https://github.com/trygve-lie/ttl-mem-cache/issues/107)) ([d7392d2](https://github.com/trygve-lie/ttl-mem-cache/commit/d7392d2a939837f77a70dc5df5f9c14ce717ed57))
* Update dependencies ([#105](https://github.com/trygve-lie/ttl-mem-cache/issues/105)) ([c9892fa](https://github.com/trygve-lie/ttl-mem-cache/commit/c9892fa3f3f25fb52bb593a145279cbe92a4b61b))


test/entry.js
'use strict';

const lolex = require('lolex');
const tap = require('tap');
const os = require('os');
const Entry = require('../lib/entry');

/**
 * Constructor
 */

tap.test('entry() - object type - should be TtlMemCacheEntry', (t) => {
    const entry = new Entry();
    t.equal(Object.prototype.toString.call(entry), '[object TtlMemCacheEntry]');
    t.end();
});

tap.test('entry() - no values - should set defaults', (t) => {
    const clock = lolex.install();
    clock.tick(1000);

    const entry = new Entry();
    t.type(entry.key, 'undefined');
    t.type(entry.value, 'null');
    t.type(entry.origin, 'null');

    t.equal(entry.ttl, -1);
    t.equal(entry.expires, 999);

    clock.uninstall();
    t.end();
});

tap.test('entry() - define all values - should set all values', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
        expires: 3000,
    });
    t.equal(entry.key, 'a');
    t.equal(entry.value, 'foo');
    t.equal(entry.origin, 'source');

    t.equal(entry.ttl, 1000);
    t.equal(entry.expires, 3000);

    t.end();
});

tap.test('entry() - define all values except expires - should set all values and calculate expires', (t) => {
    const clock = lolex.install();
    clock.tick(1000);

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
    });
    t.equal(entry.key, 'a');
    t.equal(entry.value, 'foo');
    t.equal(entry.origin, 'source');

    t.equal(entry.ttl, 1000);
    t.equal(entry.expires, 2000);

    clock.uninstall();
    t.end();
});

/**
 * Setters
 */

tap.test('entry.key - set value on "key" property - should throw', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
    });
    t.throws(() => {
        entry.key = 'b';
    }, new Error('Cannot set read-only property.'));
    t.end();
});

tap.test('entry.value - set value on "value" property - should throw', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
    });
    t.throws(() => {
        entry.value = 'bar';
    }, new Error('Cannot set read-only property.'));
    t.end();
});

tap.test('entry.ttl - set value on "ttl" property - should throw', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
    });
    t.throws(() => {
        entry.ttl = 2000;
    }, new Error('Cannot set read-only property.'));
    t.end();
});

tap.test('entry.origin - set value on "origin" property - should throw', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
    });
    t.throws(() => {
        entry.origin = 'src';
    }, new Error('Cannot set read-only property.'));
    t.end();
});

tap.test('entry.expires - set value on "expires" property - should throw', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 1000,
        origin: 'source',
    });
    t.throws(() => {
        entry.expires = 2000;
    }, new Error('Cannot set read-only property.'));
    t.end();
});

/**
 * .expired()
 */

tap.test('.expired() - expire time is in front of now - should return "false"', (t) => {
    const clock = lolex.install();

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    clock.tick(1000);

    t.notOk(entry.expired());

    clock.uninstall();
    t.end();
});

tap.test('.expired() - expire time is behind of now - should return "true"', (t) => {
    const clock = lolex.install();

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    clock.tick(4000);

    t.ok(entry.expired());

    clock.uninstall();
    t.end();
});

tap.test('.expired() - "now" argument is provided - expire time is in front of now - should return "false"', (t) => {
    const clock = lolex.install();

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    clock.tick(4000);

    t.notOk(entry.expired(1000));

    clock.uninstall();
    t.end();
});

tap.test('.expired() - "now" argument is provided - expire time is behind of now - should return "true"', (t) => {
    const clock = lolex.install();

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    clock.tick(1000);

    t.ok(entry.expired(4000));

    clock.uninstall();
    t.end();
});

/**
 * .toJSON()
 */

tap.test('.toJSON() - call method - should return an object literal of the Entry object', (t) => {
    const clock = lolex.install();
    clock.tick(1000);

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    const result = entry.toJSON();

    t.type(result, 'object');
    t.equal(result.key, 'a');
    t.equal(result.value, 'foo');
    t.equal(result.ttl, 2000);
    t.equal(result.origin, 'source');
    t.equal(result.expires, 3000);

    clock.uninstall();
    t.end();
});

tap.test('.toJSON() - call method - should append a "type" property with "TtlMemCacheEntry" as value', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    const result = entry.toJSON();

    t.equal(result.type, 'TtlMemCacheEntry');
    t.end();
});

tap.test('.toJSON() - call JSON.stringify on Entry object - should return a String representation of the Entry object', (t) => {
    const clock = lolex.install();
    clock.tick(1000);

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    const result = JSON.stringify(entry);
    t.equal(result, '{"key":"a","value":"foo","ttl":2000,"origin":"source","expires":3000,"type":"TtlMemCacheEntry"}');

    clock.uninstall();
    t.end();
});

/**
 * .assertLoose()
 */

tap.test('.assertLoose() - Entry object - should return true', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });
    t.ok(Entry.assertLoose(entry));
    t.end();
});

tap.test('.assertLoose() - Object literal with "key" and "value" - should return true', (t) => {
    const entry = {
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    };
    t.ok(Entry.assertLoose(entry));
    t.end();
});

tap.test('.assertLoose() - Object literal without "key" - should return false', (t) => {
    const entry = {
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    };
    t.notOk(Entry.assertLoose(entry));
    t.end();
});

tap.test('.assertLoose() - Object literal without "value" - should return false', (t) => {
    const entry = {
        key: 'a',
        ttl: 2000,
        origin: 'source',
    };
    t.notOk(Entry.assertLoose(entry));
    t.end();
});

tap.test('.assertLoose() - Object literal without "key" and "value" - should return false', (t) => {
    t.notOk(Entry.assertLoose({}));
    t.end();
});

tap.test('.assertLoose() - No argument - should return false', (t) => {
    t.notOk(Entry.assertLoose());
    t.end();
});

/**
 * .assertStrict()
 */

tap.test('.assertStrict() - x - should return true', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });
    t.ok(Entry.assertStrict(entry));
    t.end();
});

tap.test('.assertStrict() - Object literal with "key", "value" and "ttl" - should return true', (t) => {
    const entry = {
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    };
    t.ok(Entry.assertStrict(entry));
    t.end();
});

tap.test('.assertStrict() - Object literal without "key" - should return false', (t) => {
    const entry = {
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    };
    t.notOk(Entry.assertStrict(entry));
    t.end();
});

tap.test('.assertStrict() - Object literal without "value" - should return false', (t) => {
    const entry = {
        key: 'a',
        ttl: 2000,
        origin: 'source',
    };
    t.notOk(Entry.assertStrict(entry));
    t.end();
});

tap.test('.assertStrict() - Object literal without "ttl" - should return false', (t) => {
    const entry = {
        key: 'a',
        value: 'foo',
        origin: 'source',
    };
    t.notOk(Entry.assertStrict(entry));
    t.end();
});

tap.test('.assertStrict() - No argument - should return false', (t) => {
    t.notOk(Entry.assertStrict());
    t.end();
});

/**
 * .assertStrict()
 */

tap.test('.toPrimitive() - Call entry as String - should return JSON representation of the entry object', (t) => {
    const clock = lolex.install();
    clock.tick(1000);

    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    t.equal(`${entry}`, `{"key":"a","value":"foo","ttl":2000,"origin":"source","expires":3000,"type":"TtlMemCacheEntry"}${os.EOL}`);

    clock.uninstall();
    t.end();
});

tap.test('.toPrimitive() - Call entry as String - should have EOL at the end of String', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    const result = `${entry}`;

    t.equal(result.split('}')[1], os.EOL);
    t.end();
});

tap.test('.toPrimitive() - Call entry as a non String - should return default representation', (t) => {
    const entry = new Entry({
        key: 'a',
        value: 'foo',
        ttl: 2000,
        origin: 'source',
    });

    const result = +entry; // Call it as a number

    t.type(result, 'number'); // Return value is NaN which is a number
    t.equal(result.toString(), 'NaN');
    t.end();
});


test/utils.js
'use strict';

const lolex = require('lolex');
const tap = require('tap');
const utils = require('../lib/utils');

/**
 * .calculateExpire()
 */

tap.test('utils.calculateExpire() - "maxAge" is Infinity - should return Infinity', (t) => {
    t.equal(utils.calculateExpire(Infinity), Infinity);
    t.end();
});

tap.test('utils.calculateExpire() - "maxAge" is a number - should return now timestamp plus the number', (t) => {
    const clock = lolex.install({ now: 2000 });

    t.equal(utils.calculateExpire(2000), 4000);

    clock.uninstall();
    t.end();
});

tap.test('utils.calculateExpire() - empty argument - should return now timestamp', (t) => {
    const clock = lolex.install({ now: 2000 });

    t.equal(utils.calculateExpire(), clock.now);

    clock.uninstall();
    t.end();
});

/**
 * .expired()
 */

tap.test('utils.expired() - empty argument - should return true', (t) => {
    t.ok(utils.expired());
    t.end();
});

tap.test('utils.expired() - "expires" is Infinity - should return false', (t) => {
    const expires = Infinity;
    t.notOk(utils.expired(expires));
    t.end();
});

tap.test('utils.expired() - "expires" is behind Date.now() - should return true', (t) => {
    const expires = Date.now() - 100000;
    t.ok(utils.expired(expires));
    t.end();
});

tap.test('utils.expired() - "expires" is in front of Date.now() - should return false', (t) => {
    const expires = Date.now() + 100000;
    t.notOk(utils.expired(expires));
    t.end();
});

/**
 * .isEmpty()
 */

tap.test('utils.isEmpty() - value is not empty - should return false', (t) => {
    t.notOk(utils.isEmpty('foo'));
    t.notOk(utils.isEmpty(1));
    t.notOk(utils.isEmpty({}));
    t.notOk(utils.isEmpty([1]));
    t.end();
});

tap.test('utils.isEmpty() - value is "null" - should return true', (t) => {
    t.ok(utils.isEmpty(null));
    t.end();
});

tap.test('utils.isEmpty() - value is "undefined" - should return true', (t) => {
    t.ok(utils.isEmpty(undefined));
    t.end();
});

/**
 * .isNotEmpty()
 */

tap.test('utils.isNotEmpty() - value is not empty - should return true', (t) => {
    t.ok(utils.isNotEmpty('foo'));
    t.ok(utils.isNotEmpty(1));
    t.ok(utils.isNotEmpty({}));
    t.ok(utils.isNotEmpty([1]));
    t.end();
});

tap.test('utils.isNotEmpty() - value is "null" - should return false', (t) => {
    t.notOk(utils.isNotEmpty(null));
    t.end();
});

tap.test('utils.isNotEmpty() - value is "undefined" - should return false', (t) => {
    t.notOk(utils.isNotEmpty(undefined));
    t.end();
});

/**
 * .isFunction()
 */

tap.test('utils.isFunction() - value is not a function - should return false', (t) => {
    t.notOk(utils.isFunction());
    t.notOk(utils.isFunction('foo'));
    t.notOk(utils.isFunction(1));
    t.notOk(utils.isFunction({}));
    t.notOk(utils.isFunction([1]));
    t.end();
});

tap.test('utils.isFunction() - value is a function - should return true', (t) => {
    const fn = function fn(x) {
        return x;
    };
    const arr = (x) => {
        return x;
    };
    t.ok(utils.isFunction(fn));
    t.ok(utils.isFunction(arr));
    t.end();
});


lib/cache.js
'use strict';

/* eslint prefer-rest-params: "off" */

const stream = require('readable-stream');
const crypto = require('crypto');
const utils = require('./utils');
const Entry = require('./entry');

const _set = Symbol('_set');
const _del = Symbol('_del');

const TtlMemCache = class TtlMemCache extends stream.Duplex {
    constructor({
        ttl = 5 * 60 * 1000, stale = false, changefeed = false, id = undefined,
    } = {}) {
        super(Object.assign({
            objectMode: true
        }, ...arguments));

        Object.defineProperty(this, 'ttl', {
            value: ttl,
            enumerable: true,
        });

        Object.defineProperty(this, 'stale', {
            value: stale,
            enumerable: true,
        });

        Object.defineProperty(this, 'changefeed', {
            value: changefeed,
            enumerable: true,
        });

        Object.defineProperty(this, 'store', {
            value: new Map(),
        });

        Object.defineProperty(this, 'id', {
            value: id || crypto.randomBytes(3 * 4).toString('base64'),
            enumerable: true,
        });

        this.on('broadcast', (entry) => {
            if (this.readableFlowing) {
                if (this._readableState.objectMode) {
                    this.push(entry);
                } else {
                    this.push(Buffer.from(entry));
                }
            }
        });

        // Avoid hitting the max listeners limit when multiple
        // streams is piped into the same stream.
        this.on('pipe', () => {
            this.setMaxListeners(this.getMaxListeners() + 1);
        });

        this.on('unpipe', () => {
            this.setMaxListeners(this.getMaxListeners() - 1);
        });
    }

    /**
     * Meta
     */

    get [Symbol.toStringTag]() {
        return 'TtlMemCache';
    }

    /**
     * Private methods
     */

    [_set]({
        key, value, ttl = this.ttl, origin = this.id, expires
    }) {
        if (utils.isEmpty(key)) {
            throw new Error('Argument "key" cannot be null or undefined');
        }
        if (utils.isEmpty(value)) {
            throw new Error('Argument "value" cannot be null or undefined');
        }

        let item = value;
        if (this.changefeed) {
            item = {
                oldVal: this.get(key),
                newVal: value
            };
        }

        const entry = new Entry({
            key, value, ttl, origin, expires
        });

        this.store.set(key, entry);

        this.emit('broadcast', entry);
        this.emit('set', key, item);

        return value;
    }

    [_del]({ key }) {
        const item = this.store.get(key);
        const success = this.store.delete(key);
        if (utils.isNotEmpty(item)) {
            this.emit('broadcast', new Entry({ key }));
            this.emit('dispose', key, item.value);
        }
        return success;
    }

    /**
     * Public methods
     */

    get(key) {
        const item = this.store.get(key);
        if (utils.isNotEmpty(item)) {
            const expired = item.expired();

            if (expired) {
                this.del(key);
            }

            if (expired && this.stale) {
                return item.value;
            }

            if (expired) {
                return null;
            }

            return item.value;
        }
        return null;
    }

    set(key, value, ttl) {
        return this[_set]({ key, value, ttl });
    }

    del(key) {
        return this[_del]({ key });
    }

    entries(mutator) {
        const mutate = utils.isFunction(mutator);
        const now = Date.now();
        const arr = [];

        this.store.forEach((item, key) => {
            if (item.expired(now)) {
                if (this.stale) {
                    if (mutate) {
                        arr.push(mutator(item.value));
                    } else {
                        arr.push(item.value);
                    }
                }
                this.del(key);
            } else if (mutate) {
                arr.push(mutator(item.value));
            } else {
                arr.push(item.value);
            }
        });

        return arr;
    }

    prune() {
        const now = Date.now();
        this.store.forEach((item, key) => {
            if (item.expired(now)) {
                this.del(key);
            }
        });
    }

    clear() {
        this.store.clear();
        this.emit('clear');
    }

    dump() {
        return Array.from(this.store.entries());
    }

    load(items) {
        if (!Array.isArray(items)) {
            throw new Error('Argument "items" is not an Array');
        }
        return items.map((item) => {
            if (Entry.assertStrict(item[1])) {
                const entry = new Entry(item[1]);
                this.store.set(entry.key, entry);
                return entry.key;
            }
            return undefined;
        }).filter((item) => item);
    }

    length() {
        return this.store.size;
    }

    /**
     * Stream methods
     */

    _write(chunk, enc, next) {
        let obj = {};
        if (this._writableState.objectMode) {
            obj = chunk;
        } else {
            try {
                const str = chunk.toString();
                obj = JSON.parse(str);
            } catch (error) {
                this.emit('error', error);
                return next();
            }
        }

        if (Entry.assertLoose(obj)) {
            const item = this.store.get(obj.key);
            if (
                utils.isNotEmpty(item)
                && obj.key === item.key
                && obj.expires === item.expires
            ) {
                return next();
            }

            this[_set](obj);
            return next();
        }

        if (utils.isNotEmpty(obj.key) && utils.isEmpty(obj.value)) {
            this[_del](obj);
            return next();
        }

        this.emit('error', new Error('Object does not contain a "key" property or the value for "key" is null or undefined'));
        return next();
    }

    _read() {
        // "push" happens on "set" event in constructor
    }
};

module.exports = TtlMemCache;
